tests/infrastructure/src/main/java/org/mule/test/infrastructure/process/rules/MuleDeployment.java

```diff
@@ -192,16 +199,46 @@ protected void before() throws Throwable
         super.before();
         prober = new PollingProber(deploymentTimeout, POLL_DELAY_MILLIS);
         mule = new MuleProcessController(getMuleHome());
-        libraries.forEach((library) -> mule.addLibrary(new File(library)));
-        domains.forEach((domain) -> mule.deployDomain(domain));
-        applications.forEach((application) -> mule.deploy(application));
-        mule.start(toArray(properties));
-        logger.info("Starting Mule Server");
-        domains.forEach((domain) -> checkDomainIsDeployed(getName(domain)));
-        applications.forEach((application) -> checkAppIsDeployed(getName(application)));
-        logger.info("Deployment successful");
+        if (!mule.isRunning())
+        {
+            libraries.forEach((library) -> mule.addLibrary(new File(library)));
+            domains.forEach((domain) -> mule.deployDomain(domain));
+            applications.forEach((application) -> mule.deploy(application));
+            mule.start(toArray(properties));
+            logger.info("Starting Mule Server");
+            domains.forEach((domain) -> checkDomainIsDeployed(getName(domain)));
+            applications.forEach((application) -> checkAppIsDeployed(getName(application)));
+            logger.info("Deployment successful");
+        }
+        else
+        {
+            logger.warn("Mule Server was already running");
+            libraries.forEach((library) -> mule.addLibrary(new File(library)));
+            logger.info("Redeploying domains");
+            domains.forEach((domain) -> redeployDomain(domain));
+            logger.info("Redeploying applications");
+            applications.forEach((application) -> redeploy(application));
```

Comment:
What happens if Mule is running but I'm trying to deploy a different application (i changed its name), this is going to call **redeploy** and try to undeploy the unexisting application. Should we consider this scenario?